Move download, SHA256 check and unzip of Windows buildenv from windows_buildenv.bat to CMakelists.txt (#13306)
mixxxdj/mixxx: CMakeLists.txt

  cmake_policy(SET CMP0135 NEW)
endif()

if(WIN32)
    if(DEFINED ENV{BUILDENV_BASEPATH})

>         set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}\\")

Reviewer: In other places we ignore environment variables if they are explicit passed to cmake via -D. 
I think we should follow the same priority here and below with BUILDENV_URL

    else(NOT DEFINED BUILDENV_BASEPATH)

>         set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv/")

Reviewer: I think this is redundant with MIXXX_VCPKG_ROOT.

Reviewer: I wonder how it is done in case of git worktrees. I can imagine to have only one buildenv folder that is accessed by all worktrees. Did you consider that? That will safe a lot HD and bandwidth.

Author: Neither the variable nor the directory structure is changed, in this PR. This is therefore out of scope.
But I think it should help in this usecase, that the variable is now an environment variable, instead of a batch file internal variable. This allows the user to overwrite it after executing the script.

Author: I will replace MIXXX_VCPKG_ROOT with BUILDENV_BASEPATH, than the variable naming scheme becomes consistent.

Reviewer: It depends. I have no special demands here, just note that the VCPKG_ROOT env variable is the system wide default. We have created our own derived from that name to be independent from it.

Author: Looking at this again, it's `BUILDENV_PATH` which is redundant to `MIXXX_VCPKG_ROOT` and not `BUILDENV_BASEPATH`.

    endif()

    if(DEFINED ENV{BUILDENV_URL})
        set(BUILDENV_URL "$ENV{BUILDENV_URL}")
    elseif(NOT DEFINED BUILDENV_URL)
        message(FATAL_ERROR "BUILDENV_URL not specified")
    endif()

    # Extract the filename from the URL
    get_filename_component(FILENAME_WITH_EXTENSION "${BUILDENV_URL}" NAME)

>     # Remove the extension from the filename (note that our buildenv names contain two dots)

Reviewer: Adding an example to the comment would be helpful.

  cmake_policy(SET CMP0135 NEW)
endif()

> if(WIN32)

Reviewer 2: I know you mentioned it as out-of-scope, but I agree that sharing this logic with macOS would be great, given that the vcpkg envs are structured practically identically as far as I understand.

Author: It would be great if you could create a follow-up PR for macOS.

  cmake_policy(SET CMP0135 NEW)
endif()

> if(WIN32)

Reviewer: This should avoid reconfiguration when the CMAKE_TOOLCHAIN_FILE still exists 
```suggestion
if(WIN32 AND NOT EXISTS "${CMAKE_TOOLCHAIN_FILE}")
```

> if(WIN32)

Reviewer: ```suggestion
if(WIN32 AND (NOT EXISTS "${CMAKE_TOOLCHAIN_FILE}" OR NOT "${CMAKE_TOOLCHAIN_FILE}" MATCHES ".*${BUILDENV_NAME}.*"))
```

Author: `${CMAKE_TOOLCHAIN_FILE}` will never match `${BUILDENV_BASEPATH}/${BUILDENV_NAME}`, as the one is a file and the other a directory.
In my case `D:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-c15790e\scripts\buildsystems\vcpkg.cmake` vs `D:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-c15790e\`

What do you want to achieve with this condition?

  cmake_policy(SET CMP0135 NEW)
endif()

> if(WIN32)

Reviewer: I think we need here to check also for MIXXX_VCPKG_ROOT if that exists, we are likely in an already configured build directory and can skip the whole block. 
```suggestion
if(WIN32 AND IS_DIRECTORY MIXXX_VCPKG_ROOT)
```

Author: This would no longer detect if someone changed the buildenv to use by specifing -DBUILDENV_NAME The typical use case when you download a buildenv ZIP file from a VCPKG PR.

Reviewer: Ah I see. However it looks like id fails now when BUILDENV_NAME  and BUILDENV_URL is not even though if MIXXX_VCPKG_ROOT is valid. I consider BUILDENV_NAME and BUILDENV_URL as internal, used for communication form the batch file. Does that make sense?

Author: You mean the case that somebody locally builds VCPKG and just set MIXXX_VCPKG_ROOT ? In all other cases BUILDENV_NAME  or BUILDENV_URL are the leading vars, where MIXXX_VCPKG_ROOT could be derived from. And if windows_buildenv.bat is used, all vars are in sync anyway.

Reviewer: This is the new situation now after this PR. If one uses a branch from before only MIXXX_VCPKG_ROOT is used by cmake. "VCPKG_ROOT" is the official upstream way to point to the build environment. We have introduced MIXXX_VCPKG_ROOT for the same purpose because "VCPKG_ROOT" is already set to a different directory on the GitHub workflow runners. 

That's why I think it is more intuitive to keep "MIXX_VCPKG_ROOT" for the case you want to switch between two environments without issuing a download. I consider BUILDENV_NAME and BUILDENV_URL as programmatic script helper the user should not set manually.

Author: No it's the other way around, MIXXX_VCPKG_ROOT is just a a concatenation from BUILDENV_BASEPATH and BUILDENV_NAME inside windows_buildenv.bat. I will move this concatenation from windows_buildenv.bat to CMakeLists.txt than we have a straight forward data flow without duplicated env variables.
The basepath should be overwriteable for it's own, to allow a central storage location used by all worktrees.

      set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}")
    else()
      set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv")
    endif()
  endif()

  if(NOT DEFINED BUILDENV_URL)
    if(DEFINED ENV{BUILDENV_URL})
      set(BUILDENV_URL "$ENV{BUILDENV_URL}")
    else()
      message(STATUS "BUILDENV_URL not specified")
    endif()
  endif()

  if(DEFINED BUILDENV_URL)
    # Extract the filename from the URL
    get_filename_component(FILENAME_WITH_EXTENSION "${BUILDENV_URL}" NAME)

    # Remove the extension from the filename
    # Note, that our buildenv file names contain two dots, e.g. mixxx-deps-2.5-x64-windows-875f13b.zip
    string(
      REGEX REPLACE
      "\\.[^.]*$"
      ""
      BUILDENV_NAME
      "${FILENAME_WITH_EXTENSION}"
    )
  else()
    if(NOT DEFINED BUILDENV_NAME)
      if(DEFINED ENV{BUILDENV_NAME})
        set(BUILDENV_NAME "$ENV{BUILDENV_NAME}")
      else()
        message(
          FATAL_ERROR
          "Neither BUILDENV_URL nor BUILDENV_NAME specified!\nDid you run `${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat`?"
        )
      endif()
    endif()
  endif()

  message(STATUS "BUILDENV_NAME is: ${BUILDENV_NAME}")

>   if(
>     EXISTS "${BUILDENV_BASEPATH}/${BUILDENV_NAME}"
>     AND IS_DIRECTORY "${BUILDENV_BASEPATH}/${BUILDENV_NAME}"
>   )

Reviewer: This implies EXISTS
```suggestion
  if(IS_DIRECTORY "${BUILDENV_BASEPATH}/${BUILDENV_NAME}")
```

    message(
      STATUS

>       "BUILDENV directory ${BUILDENV_BASEPATH}/${BUILDENV_NAME} exists!"

Reviewer: Just realized even this is redundant. We have now: 
```
-- BUILDENV_NAME is: mixxx-deps-2.5-x64-windows-release-40c29ff
-- BUILDENV directory C:\buildenv/mixxx-deps-2.5-x64-windows-release-40c29ff exists!
-- Using MIXXX_VCPKG_ROOT: C:\buildenv\mixxx-deps-2.5-x64-windows-release-40c29ff
```
The last should be enough.

>       "BUILDENV directory ${BUILDENV_BASEPATH}/${BUILDENV_NAME} exists!"

Author: As this comes from different settings, I could imagine failures where both information helps to understand a problem.
It should be the same, but the information is two seperate things:
1.) The directory really exist on the harddisk
2.) How MIXXX_VCPKG_ROOT is specified
These are all existing variables, that were not introduced by this PR.

>       "BUILDENV directory ${BUILDENV_BASEPATH}/${BUILDENV_NAME} exists!"

Reviewer: We are looking now at the "good" case where we don't have a problem. Maybe there is an issue with passing redundant information? What is the expected behavior if MIXXX_VCPKG_ROOT and  BUILDENV_NAME are not matching? MIXXX_VCPKG_ROOT is probably the cached variable indicating the old sate before a vcpkg update.  

Does BUILDENV_NAME also co to CMakeCache.txt? 

How about this approach: 
* plain cmake call (without any ENV vars set) :
  * use MIXXX_VCPKG_ROOT if in cache 
  * else fail and prompt to run the batch file 
* run cmake after running the batch: 
  * Update buid environment. 

What shall happen if one changes directory to a different working directory with a different environment? 
It will be probably also updated undesired. Can we prevent that, be resetting the env variables after using?

>       "BUILDENV directory ${BUILDENV_BASEPATH}/${BUILDENV_NAME} exists!"

Author: I moved the concatenation of BUILDENV_BASEPATH + BUILDENV_NAME to MIXXX_VCPKG_ROOT from windows_buildenv.bat to CMakeLists.txt. Now we have a straightforward data flow that eliminates the possibility of contradictory information by design.

    )
  else()
    if(NOT DEFINED BUILDENV_URL)
      message(
        FATAL_ERROR
        "BUILDENV_URL not specified!\nDid you run `${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat`?"
      )
    endif()
    if(NOT DEFINED BUILDENV_SHA256)
      if(DEFINED ENV{BUILDENV_SHA256})
        set(BUILDENV_SHA256 "$ENV{BUILDENV_SHA256}")
      else()

>         message(STATUS "BUILDENV_SHA256 not specified")

Reviewer: What might have happen for this message? Can we stop here fatal?

Author: This is intentional, similar to the checksums of VCPKG ports. You can run it with out checksum specified. It will of cause fail than later, but with the actual checksum printed, that you can copy paste than.

Reviewer: It looks like the check is skipped below in this case, is it? Maybe it is cleaner to move this check closer to usage, to make it easier to trace the control flow. Do we need a bypass feature?
VCPKG has handled a missing checksum like this: 
```
  vcpkg_download_distfile requires a SHA512 argument.
  If you do not know the SHA512, add it as 'SHA512 0' and retry.
```
do we want to copy that?

Author: In the VCPKG case you need to write something into the file (works with any value btw. not only 0). But here this would be an additional step without benefit. currently it prints the information you need to specify BUILDENV_SHA256 without such an extra step. I think better developer experience is not possible here.

      endif()
    endif()

    if(NOT EXISTS "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip")
      message(
        STATUS
        "Downloading file ${BUILDENV_URL} to ${BUILDENV_BASEPATH} ..."
      )
      file(
        DOWNLOAD ${BUILDENV_URL} "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip"
        SHOW_PROGRESS
        TLS_VERIFY ON
      )
    endif()

    if(NOT ${BUILDENV_SHA256} STREQUAL "")
      message(
        STATUS
        "Verify SHA256 of downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip ..."
      )
      file(SHA256 "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip" actual_sha256)

      if(NOT actual_sha256 STREQUAL ${BUILDENV_SHA256})
        message(
          FATAL_ERROR
          "SHA256 checksum mismatch:\nexpected: ${BUILDENV_SHA256}\n     got: ${actual_sha256}"
        )
      else()

>         message(STATUS "SHA256 ${BUILDENV_SHA256} is correct!")

Reviewer: Did you consider to join the initial message with this one: 
      message(
        STATUS
        "SHA256 checksum verified for downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip"
      )

and 
      message(
          FATAL_ERROR
          "SHA256 checksum mismatch for downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip:\nexpected: ${BUILDENV_SHA256}\n     got: ${actual_sha256}\nDownload may be corrupted. Please try again."
        )

Author: I would like to leave it as it is, as the one message tells what check is started, and another with result after. I think this is easier to understand than all in one line, and the user is always aware, what happens - this is a check that need 2-3 seconds.

Reviewer: Ah OK that makes sense.

      endif()
    endif()

    message(
      STATUS
      "Unpacking file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip ..."
    )
    execute_process(
      COMMAND ${CMAKE_COMMAND} -E tar xzf "${BUILDENV_NAME}.zip"
      WORKING_DIRECTORY ${BUILDENV_BASEPATH}
      RESULT_VARIABLE unpack_result
    )
    if(NOT unpack_result EQUAL 0)
      message(FATAL_ERROR "Failed to unpack ${BUILDENV_NAME}.zip")
    else()
      file(REMOVE "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip")
      message(STATUS "Deleted ${BUILDENV_NAME}.zip after successful extraction")
    endif()
  endif()
endif()

# Use this function to throw an error because the build environment is not set
# up correctly.
function(fatal_error_missing_env)
  if(WIN32)
    if(CMAKE_BUILD_TYPE MATCHES "Debug")
      message(
        FATAL_ERROR

>         "Did you run `${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat`?"

Reviewer: The information that a GB download is started by this is missing. That can be an issue if you are out of disk space or using a slow or expensive internet connection. 
Can we add this info to the windows_buildenv.bat 

Some think like this: 

```
echo Note: 
echo This will download the Mixxx build environment which is > 1 GB.
echo Make sure you have enough disk space and a fast connection with an unlimited plan.
echo.
choice /C YN /N /M "Press Y to continue or N to cancel: "
if errorlevel 2 (
    exit /b
)
```

Author: The batch file does not download anymore, it just writes a settings file and sets some environment variables.
The main goal of this PR is to get rid of application logic in this nasty batch file. The last thing we should do is to add new functionality to this batch file.

Reviewer: This happens now "delayed". I think a typical run is like that: 
* Use cmake without the environment set (no download)  
* Read this prompt 
* Run the batch  
* Run cmake again 
* Being surprised that a GB download is started.

Before the situation was not ideal, but sightly better because only the dedicated batch call did the download, which is documented in the wiki. 

I have stumbled here because you have removed the info that a download happens. It still happens after running the script, but during the next cmake call. This can happen more or less nice by the IDE. 

Do you have an idea  how to improve the situation?

Author: I will rephrase the messages, to provide the removed information in a way that matchs to the new flow.

Author: I changed the messages and now the information about the download is back.

  cmake_policy(SET CMP0135 NEW)
endif()

> if(WIN32)
>   string(REPLACE "\\" "/" BUILDENV_BASEPATH "${BUILDENV_BASEPATH}")
>   string(REPLACE "\\" "/" CMAKE_TOOLCHAIN_FILE "${CMAKE_TOOLCHAIN_FILE}")
> endif()

Reviewer: This mirrors the CMAKE_TOOLCHAIN_FILE path handling, but uses a new variable to not change any cached value. 
```suggestion
file(
  REAL_PATH
  "${CMAKE_TOOLCHAIN_FILE}"
  CLEAN_CMAKE_TOOLCHAIN_FILE
  BASE_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
  EXPAND_TILDE
)
if(DEFINED BUILDENV_NAME)
  file(TO_CMAKE_PATH ${BUILDENV_BASEPATH} BUILDENV_BASEPATH)
  set(MIXXX_VCPKG_ROOT "${BUILDENV_BASEPATH}/${BUILDENV_NAME}")
endif()
```

if(
  WIN32
  AND
    (
      NOT DEFINED CMAKE_TOOLCHAIN_FILE
      OR
        (
          DEFINED BUILDENV_BASEPATH
          AND DEFINED BUILDENV_NAME
          AND
            NOT
              CMAKE_TOOLCHAIN_FILE
                MATCHES
                "${BUILDENV_BASEPATH}/${BUILDENV_NAME}/scripts/buildsystems/vcpkg.cmake"
        )
    )

> )

Reviewer: if(
  DEFINED CMAKE_TOOLCHAIN_FILE
  AND DEFINED MIXXX_VCPKG_ROOT
  AND
    NOT
      CLEAN_CMAKE_TOOLCHAIN_FILE
        MATCHES
        "${MIXXX_VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
)
  message(
    FATAL_ERROR
    "Build environment: '${MIXXX_VCPKG_ROOT}' does not match the environment used previously: '${CMAKE_TOOLCHAIN_FILE}'. Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory."
  )
endif()

if(WIN32 AND NOT DEFINED CMAKE_TOOLCHAIN_FILE)

> if(
>   WIN32
>   AND
>     (
>       NOT DEFINED CMAKE_TOOLCHAIN_FILE
>       OR
>         (
>           DEFINED BUILDENV_BASEPATH
>           AND DEFINED BUILDENV_NAME
>           AND
>             NOT
>               CMAKE_TOOLCHAIN_FILE
>                 MATCHES
>                 "${BUILDENV_BASEPATH}/${BUILDENV_NAME}/scripts/buildsystems/vcpkg.cmake"
>         )
>     )
> )

Reviewer: ```suggestion
if(
  DEFINED CMAKE_TOOLCHAIN_FILE
  AND DEFINED MIXXX_VCPKG_ROOT
  AND
    NOT
      CLEAN_CMAKE_TOOLCHAIN_FILE
        MATCHES
        "${MIXXX_VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
)
  message(
    FATAL_ERROR
    "Build environment: '${MIXXX_VCPKG_ROOT}' does not match the environment used previously: '${CMAKE_TOOLCHAIN_FILE}'. Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory."
  )
endif()

if(WIN32 AND NOT DEFINED CMAKE_TOOLCHAIN_FILE)
  if(NOT DEFINED BUILDENV_BASEPATH)
```

Author: The suggested condition did not work, as it disabled the buildenv download in the normal workflow (because windows_buildenv.bat sets CMAKE_TOOLCHAIN_FILE). I adjusted the condition for the error message accordingly.

  cmake_policy(SET CMP0135 NEW)
endif()

if(WIN32)

>   if(DEFINED ENV{BUILDENV_BASEPATH})
>     set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}\\")
>   else(NOT DEFINED BUILDENV_BASEPATH)
>     set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv/")
>   endif()

Reviewer: In other places we ignore environment variables if they are explicit passed to cmake via -D.
I think we should follow the same priority here and below with BUILDENV_URL

```suggestion
  if(NOT DEFINED BUILDENV_BASEPATH)
    if(DEFINED ENV{BUILDENV_BASEPATH})
      set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}\\")
    else()
      set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv/")
    endif()
  endif()
```
We may even put this repeating pattern into a helper function.

Reviewer: It turns out that there is even a CMake bug in my mininal test. CMake forgets the variable when testing for the same variable on the environment.   

```
message(STATUS "-D ${BUILDENV_BASEPATH} ENV $ENV{BUILDENV_BASEPATH}")

if(NOT DEFINED BUILDENV_BASEPATH)
  message(status "1. BUILDENV_BASEPATH is not defined.")
endif()

if(DEFINED ENV{BUILDENV_BASEPATH})
  set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}\\")
else(NOT DEFINED BUILDENV_BASEPATH)
  message(status "2. BUILDENV_BASEPATH is not defined.")
  set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv/")
endif()
  
message(FATAL_ERROR "3. ${BUILDENV_BASEPATH}")
```

```
$ cmake ../.. -DBUILDENV_BASEPATH="Hallo"
-- CMAKE_VERSION: 3.22.1
-- -D Hallo ENV 
status2. BUILDENV_BASEPATH is not defined.
CMake Error at CMakeLists.txt:64 (message):
  3.  /home/daniel/workspace/mixxx2/buildenv/
```

Reviewer: This comment is also pending

  cmake_policy(SET CMP0135 NEW)
endif()

if(WIN32 AND NOT IS_DIRECTORY MIXXX_VCPKG_ROOT)
  if(NOT DEFINED BUILDENV_BASEPATH)
    if(DEFINED ENV{BUILDENV_BASEPATH})
      set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}\\")
    else()
      set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv/")
    endif()
  endif()

>   if(DEFINED ENV{BUILDENV_URL})

Reviewer: The same here. ENV shall be used if not manually passed.

    set(BUILDENV_URL "$ENV{BUILDENV_URL}")
  elseif(NOT DEFINED BUILDENV_URL)
    message(FATAL_ERROR "BUILDENV_URL not specified")
  endif()

  # Extract the filename from the URL
  get_filename_component(FILENAME_WITH_EXTENSION "${BUILDENV_URL}" NAME)

  # Remove the extension from the filename (note that our buildenv names contain two dots)
  string(
    REGEX REPLACE
    "\\.[^.]*$"
    ""
    BUILDENV_NAME
    "${FILENAME_WITH_EXTENSION}"
  )

  message(STATUS "BUILDENV_NAME is: ${BUILDENV_NAME}")

  if(
    NOT EXISTS "${BUILDENV_BASEPATH}${BUILDENV_NAME}"
    OR NOT IS_DIRECTORY "${BUILDENV_BASEPATH}${BUILDENV_NAME}"
  )
    if(DEFINED ENV{BUILDENV_SHA256})
      set(BUILDENV_SHA256 "$ENV{BUILDENV_SHA256}")
    elseif(NOT DEFINED BUILDENV_SHA256)
      message(STATUS "BUILDENV_SHA256 not specified")

>     endif()

Reviewer: and here ...
